### PR TITLE
Ensure the existence of `DDPPlugin._sync_dir` in `reconciliate_processes`

### LINKED ISSUE #8895: Runtime events with support for custom handlers
### Motivation

In environments that deploys Lightning at a large scale, it is very useful to collect certain types of trainer runtime event into an analytics system, so engineers can monitor the overall healthiness of the Lightning usage in that environment, as well as obtain useful information for troubleshooting failed/stuck jobs.

Currently, Lightning mainly relies on logging for communicating trainer runtime information to the users. However, logging is not ideal for large scale deployments for the following reasons:
- Many training jobs in large scale deployments are automated, long-running batch jobs. It is impractical to expect the users to notice certain warnings (e.g. API deprecation warning) in a timely fashion.
- The desirable amount of events for debuggability purposes may be too verbose for logging.
- Logging is generally less structured.

### Pitch

- Introduce the concept of runtime event in Lightning
- Emit appropriate runtime events at appropriate places (e.g. deprecated API usage)
- Provide a mechanism to register custom handlers for runtime events
    - Ideally, we'd like a mechanism to transparently register default handlers (unlike `Plugin` which requires users to explicitly pass to `Trainer`), which don’t belong in the core trainer directly. This could make Lightning more attractive for larger organizations that relies on shared tooling.
    - The backend registration mechanism of `fsspec` is an ideal candidate solution: https://filesystem-spec.readthedocs.io/en/latest/developer.html#implementing-a-backend

### Additional context

______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

<sub>

- [**Metrics**](https://github.com/PyTorchLightning/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Flash**](https://github.com/PyTorchLightning/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, finetuning and solving problems with deep learning

- [**Bolts**](https://github.com/PyTorchLightning/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks and more for research and production with PyTorch Lightning and PyTorch

- [**Lightning Transformers**](https://github.com/PyTorchLightning/lightning-transformers): Flexible interface for high performance research using SOTA Transformers leveraging Pytorch Lightning, Transformers, and Hydra.

</sub>

### LINKED ISSUE #8938: `DDPPlugin.reconciliate_processes` fails on clusters that periodically purges the `/tmp`
## 🐛 Bug

`DPPPlugin` uses a temporary directory for straggler detection. If a process doesn't create a dummy file in the directory in a timely fashion upon trainer error, it is considered a straggler.

The temporary directory is created under `/tmp` (through `tempfile.mkdtemp`) in the beginning of the training. However, some clusters are configured to periodically purge the `/tmp` directory, so it may not exist anymore when it's used. When this happens, `reconciliate_processes` would fail and swallow the original error.

### To Reproduce

```python3
#!/usr/bin/env python3

import shutil
import tempfile
import time

import torch.distributed as dist

import pytorch_lightning as pl
from tests.helpers import BoringModel


class MyBoringModel(BoringModel):
    def training_step(self, batch, batch_idx):
        if dist.get_rank() == 0:
            # Simulate the periodic `/tmp` purge by the cluster
            shutil.rmtree(self.trainer.training_type_plugin._sync_dir)
            raise Exception("rank 0 failure")
        else:
            time.sleep(10)


if __name__ == "__main__":
    m = MyBoringModel()
    trainer = pl.Trainer(default_root_dir=tempfile.mkdtemp(), accelerator="ddp", num_processes=2)
    trainer.fit(m)
```

Save it as `repro.py` and run:

```bash
python3 -m torch.distributed.launch --nnodes=1 --nproc_per_node=2 repro.py
```

### Expected behavior

The straggler detection logic should be more resilient to clusters that are configured to periodically purge `/tmp`. At least it shouldn't swallow the original error.

### Environment

- PyTorch Lightning Version (e.g., 1.3.0):
- PyTorch Version (e.g., 1.8)
- Python version:
- OS (e.g., Linux):
- CUDA/cuDNN version:
- GPU models and configuration:
- How you installed PyTorch (`conda`, `pip`, source):
- If compiling from source, the output of `torch.__config__.show()`:
- Any other relevant information:

### PR DESCRIPTION
**Pull request by yifuwang**
## What does this PR do?

Fixes #8938

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Review by ananthsub — APPROVED** [review]
Thanks for the fix!

**Review by tchaton — APPROVED** [review]
Oh interesting !

**Review by kaushikb11 — APPROVED** [review]
